types/module/module.go

```diff
@@ -17,89 +17,122 @@ process. This separation is necessary, however we still want to allow for a
-	GetTxCmd() *cobra.Command
-}
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
 
 //__________________________________________________________________________________________
 // AppModuleBasic is the standard form for basic non-dependant elements of an application module.
 type AppModuleBasic interface {
 	Name() string
 	RegisterCodec(*codec.Codec)
+
+	// genesis
 	DefaultGenesis() json.RawMessage
 	ValidateGenesis(json.RawMessage) error
+
+	// client functionality
+	RegisterRESTRoutes(context.CLIContext, *mux.Router, *codec.Codec)
+	GetTxCmd() *cobra.Command
+	GetQueryCmd() *cobra.Command
 }
 
 // collections of AppModuleBasic
-type ModuleBasicManager []AppModuleBasic
+type BasicManager []AppModuleBasic
 
-func NewModuleBasicManager(modules ...AppModuleBasic) ModuleBasicManager {
+func NewBasicManager(modules ...AppModuleBasic) BasicManager {
 	return modules
 }
 
 // RegisterCodecs registers all module codecs
-func (mbm ModuleBasicManager) RegisterCodec(cdc *codec.Codec) {
-	for _, mb := range mbm {
-		mb.RegisterCodec(cdc)
+func (bm BasicManager) RegisterCodec(cdc *codec.Codec) {
+	for _, b := range bm {
+		b.RegisterCodec(cdc)
 	}
 }
 
 // Provided default genesis information for all modules
-func (mbm ModuleBasicManager) DefaultGenesis() map[string]json.RawMessage {
+func (bm BasicManager) DefaultGenesis() map[string]json.RawMessage {
 	genesis := make(map[string]json.RawMessage)
-	for _, mb := range mbm {
-		genesis[mb.Name()] = mb.DefaultGenesis()
+	for _, b := range bm {
+		genesis[b.Name()] = b.DefaultGenesis()
 	}
 	return genesis
 }
 
 // Provided default genesis information for all modules
-func (mbm ModuleBasicManager) ValidateGenesis(genesis map[string]json.RawMessage) error {
-	for _, mb := range mbm {
-		if err := mb.ValidateGenesis(genesis[mb.Name()]); err != nil {
+func (bm BasicManager) ValidateGenesis(genesis map[string]json.RawMessage) error {
+	for _, b := range bm {
+		if err := b.ValidateGenesis(genesis[b.Name()]); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
+// RegisterRestRoutes registers all module rest routes
+func (bm BasicManager) RegisterRESTRoutes(
+	ctx context.CLIContext, rtr *mux.Router, cdc *codec.Codec) {
+
+	for _, b := range bm {
+		b.RegisterRESTRoutes(ctx, rtr, cdc)
+	}
+}
+
+// add all tx commands to the rootTxCmd
+func (bm BasicManager) AddTxCommands(rootTxCmd *cobra.Command) {
+	for _, b := range bm {
+		if cmd := b.GetTxCmd(); cmd != nil {
+			rootTxCmd.AddCommand(cmd)
+		}
+	}
+}
+
+// add all query commands to the rootQueryCmd
+func (bm BasicManager) AddQueryCommands(rootQueryCmd *cobra.Command) {
+	for _, b := range bm {
+		if cmd := b.GetQueryCmd(); cmd != nil {
+			rootQueryCmd.AddCommand(cmd)
+		}
+	}
+}
+
 //_________________________________________________________
 // AppModuleGenesis is the standard form for an application module genesis functions
 type AppModuleGenesis interface {
```

Comment:
Just a quick question. So it look like `AppModuleBasic` > `AppModuleGenesis` > `AppModule`. Was there some reasoning for this structure? I feel like I might have seen this written down somewhere...

Reply:
Yeah see this explanation is at the very top of `types/module/module.go` (I just beefed it up a bit to add more description: 
```
Package module contains application module patterns and associated "manager" functionality.
The module pattern has been broken down by:
 - independent module functionality (AppModuleBasic)
 - inter-dependent module genesis functionality (AppModuleGenesis)
 - inter-dependent module full functionality (AppModule)

inter-dependent module functionality is module functionality which somehow
depends on other modules, typically through the module keeper.  Many of the
module keepers are dependent on each other, thus in order to access the full
set of module functionality we need to define all the keepers/params-store/keys
etc. This full set of advanced functionality is defined by the AppModule interface.

Independent module functions are separated to allow for the construction of the
basic application structures required early on in the application definition
and used to enable the definition of full module functionality later in the
process. This separation is necessary, however we still want to allow for a
high level pattern for modules to follow - for instance, such that we don't
have to manually register all of the codecs for all the modules. This basic
procedure as well as other basic patterns are handled through the use of
BasicManager.

Lastly the interface for genesis functionality (AppModuleGenesis) has been
separated out from full module functionality (AppModule) so that modules which
are only used for genesis can take advantage of the Module patterns without
needlessly defining many placeholder functions
```